src/gluonnlp/models/t5.py

```diff
@@ -0,0 +1,849 @@
+                dtype=dtype
+            )
+            self.cross_attn_v = nn.Dense(
+                units=self._inner_dim, 
+                in_units=d_model, 
+                flatten=False, 
+                use_bias=False, 
+                weight_initializer=Normal(d_model ** -0.5 * init_factor), 
+                dtype=dtype
+            )
+            self.cross_attn = MultiHeadAttentionCell(
+                query_units=self._inner_dim, 
+                num_heads=num_heads, 
+                attention_dropout=dropout_prob, 
+                scaled=False, 
+                normalized=False, 
+                dtype=dtype, 
+                layout='NTK' if layout == 'NT' else 'TNK', 
+                use_einsum=False
+            )
+            self.cross_attn_proj = nn.Dense(
+                units=d_model, 
+                in_units=self._inner_dim, 
+                flatten=False, 
+                use_bias=False, 
+                weight_initializer=Normal(self._inner_dim ** -0.5 * init_factor), 
+                dtype=dtype
+            )
+        self.ffn = T5FeedForward(
+            d_model=d_model, 
+            d_ff=d_ff, 
+            dropout_prob=dropout_prob, 
+            layer_norm_eps=layer_norm_eps, 
+            activation=activation, 
+            init_factor=init_factor, 
+            dtype=dtype
+        )
+        self.dropout = nn.Dropout(dropout_prob)
+
+    @property
+    def layout(self): 
+        return self._layout
+
+    @_assert_decoder_method
+    @property
+    def state_batch_axis(self): 
+        if self.layout == 'NT': 
+            return 0, 0
+        else: 
+            return 1, 1
+
+    @_assert_decoder_method
+    def _init_states(self, batch_size, ctx, dtype='float32'): 
+        if self.layout == 'NT': 
+            shape = (batch_size, 0, self._num_heads, self._d_kv)
+        else: 
+            shape = ((0, batch_size, self._num_heads, self._d_kv))
+        init_key = np.zeros(shape, ctx=ctx, dtype=dtype)
+        init_value = np.zeros(shape, ctx=ctx, dtype=dtype)
+        return init_key, init_value
+
+    @_assert_decoder_method
+    def incremental_decode(
+        self, 
+        hidden_states, 
+        past_key_value, 
+        mem_states, 
+        mem_valid_length, 
+        mem_attn_mask=None
+    ): 
+        raise NotImplementedError
+
+    def forward(
+        self, 
+        hidden_states, 
+        self_attn_mask, 
+        position_embeddings, 
+        mem_states=None, 
+        mem_attn_mask=None, 
+        mem_position_embeddings=None
+    ): 
+        """
+        hidden_states: 
+            - layout = 'NT'
+                Shape (B, L_seq, d_model)
+            - layout = 'TN'
+                Shape (L_seq, B, d_model)
+        """
+        # NT -> NTK: (B, L_seq, inner_dim) -> (B, L_seq, num_heads, n_kv)
+        # TN -> TNK: (L_seq, B, inner_dim) -> (L_seq, B, num_heads, n_kv)
+        def shape(x):
```

Comment:
Could you please rename this function to avoid ambiguity

Reply:
I kept this consistent with what huggingface has in its implementation, as I couldn't come up with a better name. Do you have any suggestion? 
FYI: https://github.com/huggingface/transformers/blob/c60e0e1ee45f4bf1017736b146c51729f120bb83/src/transformers/models/t5/modeling_t5.py#L437

Reply:
Can we refer to the name Google used in the bert source code as https://github.com/google-research/bert/blob/eedf5716ce1268e56f0a50264a88cafad334ac61/modeling.py#L629

Reply:
Sure, that sounds awesome.